### PR TITLE
feat(hydroflow_plus): use an IR before lowering to Hydroflow

### PR DESCRIPTION
**Pull request by shadaj**
feat(hydroflow_plus): use an IR before lowering to Hydroflow

Makes it possible to write custom optimization passes.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/hydro-project/hydroflow/pull/1070).
* #1080
* __->__ #1070

### CODE REVIEW

**Comment by shadaj** [comment]
Ok, this is a big one. The overall change is that rather than directly emitting Hydroflow IR, we go through a Hydroflow+ IR that captures operators through an ADT (including network across processes). This makes it dramatically easier to write optimization passes as traversals over the ADT, and helps with inspecting Hydroflow+'s view of the world (see the snapshots).

While at it, we no longer `tee()` all streams by default, instead requiring users to explicitly `.clone()` a stream to use it in multiple places. This fixes a long-standing typechecking hole in Hydroflow+ where all stream values were required to implement `Clone`. Now, tees are explicitly captured in the Hydroflow+ graph and cloning is only supported when the underlying stream elements are cloneable as well.

**Review by MingweiSamuel — APPROVED** [review]
What exactly should I think of `ir_leaves: RefCell<Vec<HfPlusLeaf>>` as

File: hydroflow_plus/src/location/mod.rs [hunk 1]
```diff
@@ -68,41 +67,23 @@ pub trait Location<'a>: Clone {
     type Meta;
 
     fn id(&self) -> usize;
-    fn flow_builder(&self) -> (&'a RefCell<usize>, &'a Builders);
+    fn flow_builder(&self) -> (&'a RefCell<usize>, &'a RefCell<Vec<HfPlusLeaf>>);
     fn next_port(&self) -> Self::Port;
 
     fn update_meta(&mut self, meta: &Self::Meta);
 
     fn spin(&self) -> Stream<'a, (), Async, Self> {
```
[the remark this answers is not in the document]
Comment (shadaj) [reply hydroflow_plus/src/location/mod.rs:75]: Yeah, will add those shortly!

File: hydroflow_plus/src/ir.rs [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+    CrossProduct(Box<HfPlusNode>, Box<HfPlusNode>),
+    Join(Box<HfPlusNode>, Box<HfPlusNode>),
+    Difference(Box<HfPlusNode>, Box<HfPlusNode>),
+    AntiJoin(Box<HfPlusNode>, Box<HfPlusNode>),
+
+    Map {
+        f: DebugExpr,
+        input: Box<HfPlusNode>,
+    },
+
```
[the remark this answers is not in the document]
Comment (shadaj) [reply hydroflow_plus/src/ir.rs:163]: Yeah basically, some can be shadowed as unknown `PipelineOp`, but for optimizations we'll want to know fine-grained details.

**Comment by shadaj** [comment]
So `ir_leaves` represents the outputs of the program (not including network between nodes), so anytime there is an external output like `for_each` or a `sink`, we emit the entire IR of the dataflow graph leading to that output. These act as roots when we emit the Hydroflow IR during compilation.

**Pull request merged**